<!-- <h1><center> <font color='blue'> IE406: Final Project </font></center></h1> -->
<h1><center> <font color='blue'>Music Genre Classification</font></center></h1>
<center><font color='blue'>Using Neural Network</font></center>

In [13]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
import keras
from keras.models import Sequential
from sklearn.model_selection import train_test_split
import numpy as np

In [14]:
DATA = pd.read_csv('Genre_data.csv')
DATA.head()

filename  length  chroma_stft_mean  chroma_stft_var  rms_mean  \
0  blues.00000.0.wav   66149          0.335406         0.091048  0.130405   
1  blues.00000.1.wav   66149          0.343065         0.086147  0.112699   
2  blues.00000.2.wav   66149          0.346815         0.092243  0.132003   
3  blues.00000.3.wav   66149          0.363639         0.086856  0.132565   
4  blues.00000.4.wav   66149          0.335579         0.088129  0.143289   

    rms_var  spectral_centroid_mean  spectral_centroid_var  \
0  0.003521             1773.065032          167541.630869   
1  0.001450             1816.693777           90525.690866   
2  0.004620             1788.539719          111407.437613   
3  0.002448             1655.289045          111952.284517   
4  0.001701             1630.656199           79667.267654   

   spectral_bandwidth_mean  spectral_bandwidth_var  ...  mfcc16_var  \
0              1972.744388           117335.771563  ...   39.687145   
1              2010.051501            65671.875673  ...   64.748276   
2              2084.565132            75124.921716  ...   67.336563   
3              1960.039988            82913.639269  ...   47.739452   
4              1948.503884            60204.020268  ...   30.336359   

   mfcc17_mean  mfcc17_var  mfcc18_mean  mfcc18_var  mfcc19_mean  mfcc19_var  \
0    -3.241280   36.488243     0.722209   38.099152    -5.050335   33.618073   
1    -6.055294   40.677654     0.159015   51.264091    -2.837699   97.030830   
2    -1.768610   28.348579     2.378768   45.717648    -1.938424   53.050835   
3    -3.841155   28.337118     1.218588   34.770935    -3.580352   50.836224   
4     0.664582   45.880913     1.689446   51.363583    -3.392489   26.738789   

   mfcc20_mean  mfcc20_var  label  
0    -0.243027   43.771767  blues  
1     5.784063   59.943081  blues  
2     2.517375   33.105122  blues  
3     3.630866   32.023678  blues  
4     0.536961   29.146694  blues  

[5 rows x 60 columns]

In [15]:
DATA.shape

(9990, 60)

In [16]:
DATA.dtypes

filename                    object
length                       int64
chroma_stft_mean           float64
chroma_stft_var            float64
rms_mean                   float64
rms_var                    float64
spectral_centroid_mean     float64
spectral_centroid_var      float64
spectral_bandwidth_mean    float64
spectral_bandwidth_var     float64
rolloff_mean               float64
rolloff_var                float64
zero_crossing_rate_mean    float64
zero_crossing_rate_var     float64
harmony_mean               float64
harmony_var                float64
perceptr_mean              float64
perceptr_var               float64
tempo                      float64
mfcc1_mean                 float64
mfcc1_var                  float64
mfcc2_mean                 float64
mfcc2_var                  float64
mfcc3_mean                 float64
mfcc3_var                  float64
mfcc4_mean                 float64
mfcc4_var                  float64
mfcc5_mean                 float64
mfcc5_var           

In [17]:
DATA = DATA.drop(['filename','length'],axis=1)

In [18]:
my_genre_list = DATA.iloc[:, -1]
le = LabelEncoder()

In [19]:
y = le.fit_transform(my_genre_list)

In [20]:
print(y)

[0 0 0 ... 9 9 9]


In [21]:
print(DATA.iloc[:, :-1])

      chroma_stft_mean  chroma_stft_var  rms_mean   rms_var  \
0             0.335406         0.091048  0.130405  0.003521   
1             0.343065         0.086147  0.112699  0.001450   
2             0.346815         0.092243  0.132003  0.004620   
3             0.363639         0.086856  0.132565  0.002448   
4             0.335579         0.088129  0.143289  0.001701   
...                ...              ...       ...       ...   
9985          0.349126         0.080515  0.050019  0.000097   
9986          0.372564         0.082626  0.057897  0.000088   
9987          0.347481         0.089019  0.052403  0.000701   
9988          0.387527         0.084815  0.066430  0.000320   
9989          0.369293         0.086759  0.050524  0.000067   

      spectral_centroid_mean  spectral_centroid_var  spectral_bandwidth_mean  \
0                1773.065032          167541.630869              1972.744388   
1                1816.693777           90525.690866              2010.051501   
2  

In [22]:
scaler = StandardScaler()
X = scaler.fit_transform(np.array(DATA.iloc[:, :-1], dtype = float))

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [24]:
# Neural network
model = Sequential()
model.add(keras.layers.Dense(256, activation='relu', input_shape=(X_train.shape[1],)))
model.add(keras.layers.Dense(128, activation='relu'))
model.add(keras.layers.Dense(64, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

In [25]:
classifier = model.fit(X_train,y_train,epochs=200,batch_size=64)

Epoch 1/200
110/110 [==============================] - 1s 2ms/step - loss: 1.6143 - accuracy: 0.4354
Epoch 2/200
110/110 [==============================] - 0s 2ms/step - loss: 0.8102 - accuracy: 0.7185
Epoch 3/200
110/110 [==============================] - 0s 2ms/step - loss: 0.6165 - accuracy: 0.7923
Epoch 4/200
110/110 [==============================] - 0s 2ms/step - loss: 0.4554 - accuracy: 0.8553
Epoch 5/200
110/110 [==============================] - 0s 2ms/step - loss: 0.3733 - accuracy: 0.8862
Epoch 6/200
110/110 [==============================] - 0s 2ms/step - loss: 0.3117 - accuracy: 0.9054
Epoch 7/200
110/110 [==============================] - 0s 2ms/step - loss: 0.2584 - accuracy: 0.9252
Epoch 8/200
110/110 [==============================] - 0s 2ms/step - loss: 0.2028 - accuracy: 0.9453
Epoch 9/200
110/110 [==============================] - 0s 2ms/step - loss: 0.1702 - accuracy: 0.9529
Epoch 10/200
110/110 [==============================] - 0s 2ms/step - loss: 0.1336 - accura

110/110 [==============================] - 0s 3ms/step - loss: 0.1243 - accuracy: 0.9580
Epoch 82/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0230 - accuracy: 0.9928
Epoch 83/200
110/110 [==============================] - 0s 3ms/step - loss: 0.0097 - accuracy: 0.9970
Epoch 84/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0043 - accuracy: 0.9988
Epoch 85/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0028 - accuracy: 0.9992
Epoch 86/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0017 - accuracy: 0.9996
Epoch 87/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0020 - accuracy: 0.9996
Epoch 88/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0017 - accuracy: 0.9996
Epoch 89/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0023 - accuracy: 0.9991
Epoch 90/200
110/110 [==============================] - 0s 2ms/step - loss: 0.0025 - accuracy: 

In [26]:
test_loss, test_acc  = model.evaluate(X_test, y_test, batch_size=64)

47/47 [==============================] - 0s 2ms/step - loss: 0.6035 - accuracy: 0.9036


## Accuracy on the training dataset is 0.9994 and on the test dataset is 0.9036